### PR TITLE
agent: Add VFIO-AP device handling

### PR DESCRIPTION
**Pull request by Jakob-Naucke**
Initial VFIO-AP support (#578) was simple, but somewhat hacky; a
different code path would be chosen for performing the hotplug, and
agent-side device handling was bound to knowing the assigned queue
numbers (APQNs) through some other means; plus the code for awaiting
them was written for the Go agent and never released. This code also
artificially increased the hotplug timeout to wait for the (relatively
expensive, thus limited to 5 seconds at the quickest) AP rescan, which
is impractical for e.g. common k8s timeouts.

Since then, the general handling logic was improved (#1190), but it
assumed PCI in several places.

In the runtime, introduce and parse AP devices. Annotate them as such
when passing to the agent, and information about the enabled APQNs is
included.

The agent awaits the passed APQNs through uevents and triggers a
rescan directly.

Fixes: #3678

### CODE REVIEW

**Comment by c3d** [comment]
@Jakob-Naucke 
> I have decided to prepend the agent device options with pci: or ap: respectively. If you strongly disagre, we can find another way.

Does not sound like a bad plan, but isn't it the case that the host pretty much determines the kind of device you have? In other words, are there platforms with both PCI and AP?

File: src/agent/src/ap.rs
Comment (Jakob-Naucke) [reply src/agent/src/ap.rs:40]: Hmm. Elsewhere, we are doing things like: https://github.com/kata-containers/kata-containers/blob/a3b3274121f4bfb0a9d0ba0759d4d2e2a778134b/src/agent/src/pci.rs#L125 or https://github.com/kata-containers/kata-containers/blob/a3b3274121f4bfb0a9d0ba0759d4d2e2a778134b/src/agent/src/ccw.rs#L54 (that one was me, to be fair). What I had wanted to express with this format is make it clear that we're expecting hex numbers here. We can also put semantic names in the error message though. WDYT?

File: src/agent/src/ap.rs
Comment (jodh-intel) [inline src/agent/src/ap.rs:73]: A couple of ideas for additional tests:

- Empty string.
- "`.`"
- Value that doesn't contain any periods.
- Value that is too big for the type.
- Non-hex numeric value.

**Comment by Jakob-Naucke** [comment]
> > I have decided to prepend the agent device options with pci: or ap: respectively. If you strongly disagre, we can find another way.
> 
> Does not sound like a bad plan, but isn't it the case that the host pretty much determines the kind of device you have? In other words, are there platforms with both PCI and AP?

AP is for s390x, which (with some specialities) also has PCI support, so yes, but those would not be within one device, so I wouldn't really see the issue.

File: src/agent/src/device.rs
Comment (jodh-intel) [inline src/agent/src/device.rs:859]: You're doing `+=` so why the `+ 5` at the end? I'm guessing that was the original value before you added `AP_CONFIG_TIME_SECONDS`?

File: src/agent/src/device.rs
Comment (jodh-intel) [inline src/agent/src/device.rs:861]: I'm confused. You changed the config on the previous line, but are now dropping it? Should that `drop()` actually happen after the `write()`?

File: src/agent/src/device.rs
Comment (jodh-intel) [inline src/agent/src/device.rs:862]: Should this actually be?:

```suggestion
        fs::write(AP_CONFIG_TIME_PATH, config.hotplug_timeout.to_string())?;
```

File: src/agent/src/device.rs
Comment (jodh-intel) [inline src/agent/src/device.rs:868]: Should this be `config.hotplug_timout` after the increment but before the `drop()`?

File: src/runtime/virtcontainers/device/config/config.go
Comment (jodh-intel) [inline src/runtime/virtcontainers/device/config/config.go:273]: If this is `true`, `IsPCIe` will also be `true` right? Is this being `true` but `IsPCIe` false a legal state (denoting AP + PCIe?) I'm wondering if these values would be better as an `enum`?

File: src/runtime/virtcontainers/kata_agent.go
Comment (jodh-intel) [inline src/runtime/virtcontainers/kata_agent.go:1224]: This suggests "AP" trumps "PCI" right?

File: src/runtime/virtcontainers/device/config/config.go
Comment (jodh-intel) [inline src/runtime/virtcontainers/device/config/config.go:272]: Is this comment correct? It seems a bit odd that there are two `bool`'s to denote a PCI device.

**Review by jodh-intel — COMMENTED** [review]
Thanks @Jakob-Naucke - a few comments...

File: src/agent/src/device.rs
Comment (Jakob-Naucke) [reply src/agent/src/device.rs:859]: Oops, that is not necessary and was added in development in confusion over what turned out to be another bug. Will remove.

File: src/agent/src/device.rs
Comment (Jakob-Naucke) [reply src/agent/src/device.rs:862]: No :) Let me explain:
- Hotplugging, generally, can experience some delays. This is the normal `hotplug_timeout`, which defaults to 3 seconds.
- The uevent for AP devices will come in upon a rescan, which happens periodically.
- This period is configurable, down to a minimum of 5 seconds, and is configured in this line.

The `hotplug_timeout` that I want to use for AP is the "general" grace period of 3 seconds plus the up-to-5 seconds until rescan.

File: src/agent/src/device.rs
Comment (Jakob-Naucke) [reply src/agent/src/device.rs:868]: No, this is just the `hotplug_timeout` as it was before we updated it. This is a regular backup-and-restore.

File: src/runtime/virtcontainers/device/config/config.go
Comment (Jakob-Naucke) [reply src/runtime/virtcontainers/device/config/config.go:273]: AP + PCIe would, indeed, not be legal. Let me look into it.

File: src/agent/src/device.rs
Comment (jodh-intel) [reply src/agent/src/device.rs:861]: Ah - it's the `lazy_static` global - ignore that comment as I'd read it as a local variable so it looked very odd to set some state, then dtrop it ;)

File: src/runtime/virtcontainers/kata_agent.go
Comment (Jakob-Naucke) [reply src/runtime/virtcontainers/kata_agent.go:1224]: Because of the order? I had just used that order to have
```
if cond {
  do_cond_stuff()
} else {
  do_non_cond_stuff()
}
```

instead of
```
if !cond {
  do_non_cond_stuff()
} else {
  do_cond_stuff()
}
```

but this would change with the enum anyway

File: src/runtime/virtcontainers/device/config/config.go
Comment (Jakob-Naucke) [reply src/runtime/virtcontainers/device/config/config.go:272]: related to AP/PCI/PCIe discussion above

File: src/agent/src/ap.rs
Comment (dgibson) [inline src/agent/src/ap.rs:19]: AFAICT this is just encoding a device address, so I wonder if `Address` might be a better name, especially with the analogy to `crate::pci::Address`.

File: src/agent/src/ap.rs
Comment (dgibson) [reply src/agent/src/ap.rs:73]: Seconded.  The unit testss in the `crate::pci` module may give some inspiration.

File: src/agent/src/device.rs
Comment (dgibson) [inline src/agent/src/device.rs:421]: This is analogous to `PciMatcher` *not* to `VfioMatcher` which you renamed to `VfioPciMatcher`.

File: src/agent/src/device.rs
Comment (dgibson) [inline src/agent/src/device.rs:440]: I do wonder if we should make a common matcher which just checks syspath, for all the cases where we can determine the exact sysfs path ahead of time.  Then we can just have multiple constructors or helper functions to generate that matcher based on whatever info we have.

That can be a clean up for another time though.

File: src/agent/src/device.rs
Comment (dgibson) [inline src/agent/src/device.rs:763]: Demultiplexing here seems silly to me.  Why not just create new device type strings for AP devices (and rename the existing PCI ones if we want)?

File: src/agent/src/device.rs
Comment (dgibson) [inline src/agent/src/device.rs:769]: Note that using `split_once` introduces a new Rust version dependency.

File: src/agent/src/device.rs
Comment (dgibson) [inline src/agent/src/device.rs:800]: Uh... doesn't the fact that you're just changing `split_vfio_option` here mean you broke compile wtih the earlier patch that renamed the function?

File: src/agent/src/device.rs
Comment (dgibson) [reply src/agent/src/device.rs:862]: Yikes.  Making the global config writable just for this seems like a terrible idea to me.  That seems really likely to break people's assumptions that it's essentially a write-once global.

File: src/agent/src/device.rs
Comment (dgibson) [reply src/agent/src/device.rs:865]: I suspect this will not work quite the way you want, or at least, not reliably.  Part of the reason for 3s timeout is that there's often a ~5s timeout imposed by the layers above (e.g. CRI-O and/or k8s).  Making sure we time out first means we can give a useful error, rather than getting just a generic "container start up timed out" error.

But the point is, that increasing this timeout won't save you: if the hotplug takes too long, you'll still hit the upper levels timeout, just with a less meaningful error.  I've hit this in practice when I attempted to increase the timeout to deal with SHPC PCI hotplug (which also has a 5s delay).

Basically if yout hotplug takes this long, it's kind of fundamentally incompatible with container usage :(.  You may need to invent a different (VM only) hotplug mechanism that will complete more quickly.

File: src/runtime/virtcontainers/device/config/config.go
Comment (dgibson) [inline src/runtime/virtcontainers/device/config/config.go:266]: What "in the mdev" means is unclear to me.  It's also unclear why they'd be mutually exclusive with PCI - s390x can have a PCI bus, IIUC, even if it's not particularly common.

File: src/runtime/virtcontainers/device/config/config.go
Comment (dgibson) [inline src/runtime/virtcontainers/device/config/config.go:267]: Why multiple devices here?  For the PCI, `VFIODev` represents a single PCI device, with multiple devices in the VFIO group being represented by multiple `VFIODev`s in a single `VFIODevice` (yes, the names are confusing).  Shouldn't the same model apply here?

Or, better yet `VFIODevice` could list either a bunch of `PCIDev`s or a bunch of `APDev`s, improving the naming and reducing redundancy in the `VFIODev` structure.

File: src/runtime/virtcontainers/kata_agent.go
Comment (dgibson) [reply src/runtime/virtcontainers/kata_agent.go:1224]: As I noted in the agent code, I think using a different `kataDevice.Type` value would make more sense than putting this into all the options.  You're also making things unnecessarily different from the PCI case by folding all the AP devices into a single option, whereas PCI devices are placed one per option.

File: src/runtime/virtcontainers/qemu.go
Comment (dgibson) [inline src/runtime/virtcontainers/qemu.go:1573]: This is the main consumer of the type value.  So it seems like it might make more sense to make AP device a separate option at that level, rather than testing for mediated type *then* for `isAP`.

**Review by dgibson — CHANGES_REQUESTED** [review]

File: src/agent/src/ap.rs
Comment (jodh-intel) [inline src/agent/src/ap.rs:12]: fyi, I harvested this for the [CC glossary](https://github.com/confidential-containers/documentation/wiki/Glossary#adjunct-processor) :smile:

File: src/agent/src/device.rs
Comment (Jakob-Naucke) [reply src/agent/src/device.rs:865]: @dgibson Thank you for your review. You made some important comments, but let me look into this before anything else because it's a potential showstopper.

**Comment by Jakob-Naucke** [comment]
I'll put a comment at the end for readability: Because of the concerns raised in https://github.com/kata-containers/kata-containers/pull/3679#discussion_r807525412, this PR is on hold until we find a better solution than needing to wait 5 seconds for the related uevent. This seems possible and we are looking into this.

File: src/agent/src/device.rs
Comment (Jakob-Naucke) [reply src/agent/src/device.rs:865]: I have a kernel patch that resolves this issue from a colleague specialised in this at this point, but let's conclude :arrow_down: first

File: src/runtime/virtcontainers/qemu.go
Comment (Jakob-Naucke) [reply src/runtime/virtcontainers/qemu.go:1573]: You make some valid concerns about API cleanliness, but there is a catch I'd like to hear your opinion on. 
@dgibson, you suggest breaking this down by `config.VFIODev.Type`, which, until now, can be a `VFIODeviceNormalType` or a `VFIODeviceMediatedType`, i.e. we'd have e.g. a `VFIOAPDeviceMediatedType` (is always mediated) (and perhaps rename the others).
Meanwhile, IIUC, both you and @jodh-intel suggest breaking down `config.VFIODev` (which also assumes PCI through and through) into two devices. (James, you said `enum` -- I guess you meant an `interface`?)
However, implementing both of these would introduce redundancy and possible illegal states (e.g. a `config.VFIOAPDev` whose `Type` is `VfioPCIDeviceNormalType` or such).
And we cannot break up the VFIO devices "from the beginning" (i.e. `DeviceManager.NewDevice`), because only after checking IOMMU groups can you know whether a device is PCI or AP. 
WDYT?

(Note that this is not the distinction of `kataDevice.Type`, which currently has `vfio` or `vfio-gk` (where `gk` is guest kernel, i.e. not for the container, which changes the agent-side handling, which I shouldn't think we'd need for AP). I guess adding an option there is not a huge conflict though.)

**Review by Jakob-Naucke — COMMENTED** [review]
@dgibson @jodh-intel please refer mostly to the longest comment :)

File: src/agent/src/device.rs
Comment (dgibson) [reply src/agent/src/device.rs:68]: Right, I was reading the commits in order, so it wasn't obvious when I wrote this.

File: src/agent/src/ap.rs
Comment (dgibson) [reply src/agent/src/ap.rs:19]: Um.. I don't really understand the question, sorry.

File: src/runtime/virtcontainers/qemu.go
Comment (dgibson) [reply src/runtime/virtcontainers/qemu.go:1573]: The Go code, and the agent-protocol is already full of places which have representable invalid states.  One more doesn't make much difference.

I doubt you need a `vfio-gk` equivalent for AP.  `vfio` passes through devices in the way that a non-Kata container would do so: a VFIO device in the host becomes a VFIO device in the guest.  `vfio-gk` is a effectively poorly-defined Kata extension, which has the VFIO device in the host become.. something.. in the guest depending what driver is available in the guest kernel.

**Review by Jakob-Naucke — COMMENTED** [review]
@jodh-intel @dgibson sorry for the wait. I _think_ this is technically what you asked for, but I can't say I'm very happy with it. Please tell me what you think.

**Review by dgibson — COMMENTED** [review]
I like this revision much more than the earlier one.

I'm now pretty much happy with the agent side code, apart from some minor things.  I do still have some thoughts on the organization of the runtime side code.

**Review by Jakob-Naucke — COMMENTED** [review]
And here's to more updates @jodh-intel @dgibson. I tried to split them out as well as I could.

**Comment by Jakob-Naucke** [comment]
Needed a `govet` update -- rebased

**Comment by Jakob-Naucke** [comment]
needed test fix -- rebased

**Comment by Jakob-Naucke** [comment]
@dgibson ping for another look if you can find the time :)

File: src/runtime/virtcontainers/device/drivers/vfio.go
Comment (dgibson) [reply src/runtime/virtcontainers/device/drivers/vfio.go:214]: New version is better.  Ugly, but that's a result of dealing with the ugly existing persist ingerface.

File: src/agent/src/ap.rs
Comment (dgibson) [reply src/agent/src/ap.rs:1]: This doesn't seem to have been addressed - I still find the commit message pretty hard to read.

File: src/agent/src/device.rs
Comment (dgibson) [reply src/agent/src/device.rs:831]: Pretty dire: in some cases it could put a PCI device in an unusable state that required a *host* reboot to fix :(.  AP is very different of course, so it wouldn't suffer the same problems technically, but like PCI this is probably a rarely tested path.

File: src/runtime/virtcontainers/device/config/config.go
Comment (dgibson) [reply src/runtime/virtcontainers/device/config/config.go:299]: Right, but AFAICT the mediated VFIO device should map to the `VFIODevice` which can already list multiple `VFIODev`s, each oone can be a single APQN.  This doesn't seem different to me from the PCI case where a single `/dev/vfio/NNN` (a single `VFIODevice`) corresponds to multiple hardware devices in the VFIO (each a `VFIODev`).  This can happen, for example, if you have a bunch of devices beneath a PCIe to PCI bridge.

**Review by dgibson — CHANGES_REQUESTED** [review]

File: src/runtime/virtcontainers/device/config/config.go
Comment (Jakob-Naucke) [reply src/runtime/virtcontainers/device/config/config.go:308]: We go through `/sys/kernel/iommu_groups/NNN/devices`, but with VFIO-AP, that is just going to be one UUID, potentially with several APQNs behind it (see also https://docs.kernel.org/s390/vfio-ap.html). I think what you're describing applies to PCI normal, but not mediated.

**Review by Jakob-Naucke — COMMENTED** [review]
@dgibson thanks for the iteration -- please see the comments below.

File: src/runtime/virtcontainers/clh.go
Comment (Jakob-Naucke) [reply src/runtime/virtcontainers/clh.go]: Oops, that one wasn't in the update previously.

File: tools/packaging/kernel/patches/5.15.x/0001-s390-ap-enable-sysfs-attribute-scans-to-force-AP-bus.patch
Comment (hbrueckner) [reply tools/packaging/kernel/patches/5.15.x/0001-s390-ap-enable-sysfs-attribute-scans-to-force-AP-bus.patch:4]: > Hm... ok. I don't love merging this in Kata before the change is merged in a released kernel upstream. Red Hat, for one, deploys Kata with a packaged guest kernel rather than something custom built with these patches.

This patch has been merged into 5.17-rc3 (https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/drivers/s390/crypto/ap_bus.c?id=8944d05f9bbf910c8b241e29a3de114900e31e42)

File: src/runtime/virtcontainers/device/drivers/vfio.go
Comment (dgibson) [reply src/runtime/virtcontainers/device/drivers/vfio.go:251]: Right, using `filepath.Base` within the function is a fine approach.  I do suggest renaming the parameter to match as well, though.

File: src/runtime/virtcontainers/device/drivers/utils.go
Comment (dgibson) [reply src/runtime/virtcontainers/device/drivers/utils.go]: Oof... ok.  Folding the error into the normal return type seems horrible; very against normal Go conventions.

File: src/runtime/virtcontainers/device/drivers/vfio.go
Comment (Jakob-Naucke) [reply src/runtime/virtcontainers/device/drivers/vfio.go:251]: Done.

File: tools/packaging/kernel/patches/5.15.x/0001-s390-ap-enable-sysfs-attribute-scans-to-force-AP-bus.patch
Comment (Jakob-Naucke) [reply tools/packaging/kernel/patches/5.15.x/0001-s390-ap-enable-sysfs-attribute-scans-to-force-AP-bus.patch]: Thanks @hbrueckner -- the included patch is now as released on kernel.org.

**Comment by arronwy** [comment]
Can one of the admins verify this patch?

**Comment by BbolroC** [comment]
Hi, I was wondering why this PR has been stuck quite a while. I would be grateful if someone from the reviewers could mention a reason and what is needed to be done in order to get this going forward. @Jakob-Naucke if you know why, please leave any meaningful feedback. Thanks.

**Comment by Jakob-Naucke** [comment]
@BbolroC hmm, there were lots of good comments made by @dgibson who, like me, isn't really active here any more. Some things weren't settled 100% -- this would need further reviewer input.

I could resolve the conflicts, but I couldn't really test the updated version. I could sit down with you briefly to assist if you have trouble and when both of us find the time.

File: src/agent/src/ap.rs
Comment (c3d) [reply src/agent/src/ap.rs:40]: Can you express the "hex" with an example?

Comment (BbolroC) [reply src/agent/src/ap.rs:40]: You mean "e.g. 0a.0033" after "<xx>.<xxxx>", right? I could update this after this gets merged.

**Comment by BbolroC** [comment]
handed over to @BbolroC

**Pull request closed without merging**
